src/estimators/exact_set.py

```diff
@@ -67,24 +85,47 @@ def __init__(self):
     EstimatorBase.__init__(self)
 
   def __call__(self, sketch_list):
-    """Return len(sketch)."""
-    union = ExactSet()
+    """Returns sketch frequency histogram."""
+    if len(sketch_list) == 0:
+      return [0]
+    union = ExactMultiSet()
     for s in sketch_list:
-      union.add_ids(s.ids())
-    return [len(union)]
+      for id in s.ids():
+        union.add_ids([id] * s.frequency(id))
+    frequency_counts = collections.defaultdict(int)
+    for x in union.ids():
+      frequency_counts[union.frequency(x)] += 1
+    # At this point, frequency_counts is a dictionary whose keys are frequencies
+    # and whose values are the number of items with the corresponding frequency.
+    # We now need to create a histogram, where histogram[i] consists of all
+    # items having frequency i or greater.  So, if frequency[i] is the number of
+    # items having exactly frequency i, we want to compute:
+    #   histogram[i] = frequency[i] + frequency[i+1] + ...
+    # To accomplish this, we convert the dictionary of frequency counts to a list,
+    # and then compute the reversed cumulative sums.
+    histogram = list(reversed(np.cumsum([frequency_counts[i] for i in range(max(frequency_counts), 0, -1)])))
+    return histogram
 
 
 class LessOneEstimator(EstimatorBase):
-  """An estimator for ExactSet that is always short by one."""
+  """Reports a cardinality one smaller than the true cardinality."""
 
   def __init__(self):
     EstimatorBase.__init__(self)
 
   def __call__(self, sketch_list):
-    """Return len(sketch) - 1."""
+    """Returns frequency histogram of sketch with total cardinality reduced by one."""
     e = LosslessEstimator()
     histogram = e(sketch_list)
-    return [histogram[0]-1] + [histogram[1:]]
+    if sum(histogram) == 0:
+      raise ValueError("Attempt to create a histogram with a negative value!")
+    # It seems pretty clear that if the original histogram is [3, 2, 1], then
+    # a reasonable output is [2, 2, 1].  This corresponds to removing (the) one item
+    # whose frequency is 1.  But what if the histogram is [2, 2, 1]?  This represents
+    # a set having 0 items of frequency 1, 1 of frequency 2, and 1 of frequency 3.
+    # What should be removed in this case, and what should the resulting histogram
+    # look like?  I have opted for removing an item of lowest positive frequency.
```

Comment:
This seems too clever/complex. Maybe just subtract one from each non-zero frequency?

The whole point of this class is really just as a test utility anyway.